### PR TITLE
Fail tests on failed assert

### PR DESCRIPTION
**Pull request by jasonmalinowski**
*NOTE:* you'll see if I've added a commit to this PR that adds a test that will fail. This is to verify it actually works on the build machines. Assuming we get a test run that fails with just that, I'll elide the commit, wait until it's green, and merge it.

Review: @dotnet/roslyn-compiler, @dotnet/roslyn-infrastructure, @jaredpar, @davkean

### CODE REVIEW

**Comment by davkean** [comment]
I think'll want to eventually centralize this - but it's a great start.

File: src/Test/Utilities/Desktop/ThrowingTraceListener.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis
+{
+    public sealed class ThrowingTraceListener : System.Diagnostics.TraceListener
```
Comment (jaredpar) [inline src/Test/Utilities/Desktop/ThrowingTraceListener.cs]: Can it be just `TraceListener`?

File: src/Test/Utilities/Desktop/ThrowingTraceListener.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        public override void Fail(string message, string detailMessage)
+        {
+            throw new DebugAssertFailureException(string.Format("{0}\r\n{1}", message, detailMessage));
```
Comment (jaredpar) [inline src/Test/Utilities/Desktop/ThrowingTraceListener.cs]: Why not `Environment.NewLine` and / or interpolated strings?

**Comment by jaredpar** [comment]
@davkean what do you mean by centralizing?  Are you envisioning a single app.config that we somehow include / copy on builds? 

@jasonmalinowski :+1:

**Comment by davkean** [comment]
Yes, have a single App.Config that we automatically share between tests, and then individual tests can either override it, or have their app.config get merged in with the common one.

File: src/Test/Utilities/Desktop/ThrowingTraceListener.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/Test/Utilities/Desktop/ThrowingTraceListener.cs]: This was just the code that was already there -- I renamed the file and Git isn't picking up on it. I can update though if you'd like.

File: src/Test/Utilities/Desktop/ThrowingTraceListener.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Test/Utilities/Desktop/ThrowingTraceListener.cs]: We could; I'd be terrified that somebody will inherit from 'TraceListener' and then inherit from this by accident.

Comment (davkean) [inline src/Test/Utilities/Desktop/ThrowingTraceListener.cs]: I think he means remove the namespace prefix.

Comment (jaredpar) [inline src/Test/Utilities/Desktop/ThrowingTraceListener.cs]: @davkean is correct

File: src/Test/Utilities/Desktop/TraceListener.cs [hunk 3]
```diff
@@ -1,81 +0,0 @@
-﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
-
-using System;
-using System.Diagnostics;
-
-namespace Microsoft.CodeAnalysis
-{
-#if DEBUG
-    // To enable, add to <listeners> in *.exe.config, specifying the assembly-qualified
```
Comment (AlekseyTs) [inline src/Test/Utilities/Desktop/TraceListener.cs]: It would be good to keep this comment explaining how one can take advantage of this type.

Comment (AlekseyTs) [inline src/Test/Utilities/Desktop/TraceListener.cs]: LGTM

**Comment by davkean** [comment]
(Closing/opening to sync update the merge branch to pick up the build break we just fixed)

**Comment by davkean** [comment]
@dotnet-bot retest prtest/lin/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/2670/
// Retest reason:

**Comment by davkean** [comment]
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/2710/
// Retest reason:

**Comment by davkean** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/2684/
// Retest reason:

**Comment by jasonmalinowski** [comment]
@dotnet-bot, retry mac please. The previous failure was a GitHub timeout when syncing.

**Comment by jasonmalinowski** [comment]
@dotnet-bot, retest mac please. The previous failure was a GitHub timeout syncing.

**Pull request merged**